sql: tuple mixing ints and floats causing panic

... and even when it doesn't crash, I think using an index when a type conversion is involved is completely broken, since the literals don't have their types inferred early enough and are encoded with the wrong type in the index spans. Unless I'm missing something?

```
CREATE TABLE a2 (a FLOAT PRIMARY KEY)
INSERT INTO a2 VALUES (2.1)
SELECT * FROM a2 WHERE a IN (1.1, 2)
```

```
panic: unsupported comparison: int to float

goroutine 8596 [running]:
github.com/cockroachdb/cockroach/sql/parser.DInt.Compare(0x2, 0x6d661b0, 0xc8203eac10, 0x0)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/sql/parser/datum.go:153 +0x2b8
github.com/cockroachdb/cockroach/sql/parser.(*DInt).Compare(0xc8203eac20, 0x6d661b0, 0xc8203eac10, 0x58f9fa0)
	<autogenerated>:193 +0xa3
github.com/cockroachdb/cockroach/sql/parser.DTuple.Less(0xc8205a7760, 0x2, 0x2, 0x1, 0x0, 0x4f225d0)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/sql/parser/datum.go:558 +0x77
github.com/cockroachdb/cockroach/sql/parser.(*DTuple).Less(0xc8205a77a0, 0x1, 0x0, 0x4fb0b60)
	<autogenerated>:148 +0xb5
sort.insertionSort(0x6db6cb0, 0xc8205a77a0, 0x0, 0x2)
	/usr/local/go/src/sort/sort.go:32 +0x6f
sort.quickSort(0x6db6cb0, 0xc8205a77a0, 0x0, 0x2, 0x4)
	/usr/local/go/src/sort/sort.go:184 +0x153
sort.Sort(0x6db6cb0, 0xc8205a77a0)
	/usr/local/go/src/sort/sort.go:199 +0x74
github.com/cockroachdb/cockroach/sql/parser.(*DTuple).Normalize(0xc8205a77a0)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/sql/parser/datum.go:567 +0x48
github.com/cockroachdb/cockroach/sql/parser.(*ComparisonExpr).normalize(0xc8200fcb40, 0xc8201ae190, 0x0, 0x0)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/sql/parser/normalize.go:195 +0xd0d
github.com/cockroachdb/cockroach/sql/parser.(*normalizeVisitor).Visit(0xc8201ae190, 0x6d66dc8, 0xc8200fcb40, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/sql/parser/normalize.go:330 +0x119
github.com/cockroachdb/cockroach/sql/parser.WalkExpr(0x6d66f40, 0xc8201ae190, 0x6d66dc8, 0xc8200fcb40, 0x0, 0x0)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/sql/parser/walk.go:214 +0xce
github.com/cockroachdb/cockroach/sql/parser.(*Parser).NormalizeExpr(0xc8201aa0e8, 0x1, 0xece1e7385, 0x1382df66, 0x58cb600, 0xece1e7385, 0x1382df66, 0x58cb600, 0xc8200ff190, 0xc8204d05a0, ...)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/sql/parser/parse.go:88 +0x102
github.com/cockroachdb/cockroach/sql.(*scanNode).initWhere(0xc8206523c0, 0xc8205a7320, 0x0, 0x0)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/sql/scan.go:375 +0x2c6
github.com/cockroachdb/cockroach/sql.(*planner).selectWithScan(0xc8201aa000, 0xc8206523c0, 0xc8205a40a0, 0x0, 0x0, 0x0, 0x0)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/sql/select.go:51 +0xfe
github.com/cockroachdb/cockroach/sql.(*planner).Select(0xc8201aa000, 0xc8205a40a0, 0x0, 0x0, 0x0, 0x0)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/sql/select.go:43 +0xaf
github.com/cockroachdb/cockroach/sql.(*planner).makePlan(0xc8201aa000, 0x6d66e40, 0xc8205a40a0, 0x0, 0x0, 0x0, 0x0)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/sql/plan.go:124 +0x136a
github.com/cockroachdb/cockroach/sql.(*Executor).execStmt.func1(0xece1e7385, 0xc81382df66, 0x58cb600, 0x0, 0x0)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/sql/executor.go:310 +0xd9
github.com/cockroachdb/cockroach/sql.(*Executor).execStmt.func2(0xc8201d4c60, 0x0, 0x0)
	/Users/andrei/work/src/github.com/cockroachdb/cockroach/sql/executor.go:386 +0x12e
```

--- a/sql/parser/datum.go
+++ b/sql/parser/datum.go
@@ -86,6 +86,7 @@ type Datum interface {
 	TypeEqual(other Datum) bool
 	// Compare returns -1 if the receiver is less than other, 0 if receiver is
 	// equal to other and +1 if receiver is greater than other.
+	// TODO(nvanbenschoten) Should we look into merging this with cmpOps?
 	Compare(other Datum) int
 	// HasPrev specifies if Prev() can be used to compute a previous value for
 	// a datum. For example, DBytes doesn't support it (the previous for BB is BAZZZ..).
@@ -211,7 +212,11 @@ func (d DInt) Compare(other Datum) int {
 	}
 	v, ok := other.(DInt)
 	if !ok {
-		panic(fmt.Sprintf("unsupported comparison: %s to %s", d.Type(), other.Type()))
+		cmp, ok := mixedTypeCompare(d, other)
+		if !ok {
+			panic(fmt.Sprintf("unsupported comparison: %s to %s", d.Type(), other.Type()))
+		}
+		return cmp
 	}
 	if d < v {
 		return -1
@@ -278,7 +283,11 @@ func (d DFloat) Compare(other Datum) int {
 	}
 	v, ok := other.(DFloat)
 	if !ok {
-		panic(fmt.Sprintf("unsupported comparison: %s to %s", d.Type(), other.Type()))
+		cmp, ok := mixedTypeCompare(d, other)
+		if !ok {
+			panic(fmt.Sprintf("unsupported comparison: %s to %s", d.Type(), other.Type()))
+		}
+		return cmp
 	}
 	if d < v {
 		return -1
@@ -363,7 +372,11 @@ func (d *DDecimal) Compare(other Datum) int {
 	}
 	v, ok := other.(*DDecimal)
 	if !ok {
-		panic(fmt.Sprintf("unsupported comparison: %s to %s", d.Type(), other.Type()))
+		cmp, ok := mixedTypeCompare(d, other)
+		if !ok {
+			panic(fmt.Sprintf("unsupported comparison: %s to %s", d.Type(), other.Type()))
+		}
+		return cmp
 	}
 	return d.Cmp(&v.Dec)
 }
@@ -1015,4 +1028,39 @@ func (DValArg) IsMin() bool {
 
 func (d DValArg) String() string {
 	return "$" + d.name
+}
+
+// Temporary workaround for #3633, allowing comparisons between
+// heterogeneous types.
+// TODO(andreimatei) Remove when type inference improves.
+func mixedTypeCompare(l, r Datum) (int, bool) {
+	lType := reflect.TypeOf(l)
+	rType := reflect.TypeOf(r)
+
+	// Check equality.
+	eqOp, ok := cmpOps[cmpArgs{EQ, lType, rType}]
+	if !ok {
+		return 0, false
+	}
+	eq, err := eqOp.fn(EvalContext{}, l, r)
+	if err != nil {
+		panic(err)
+	}
+	if eq {
+		return 0, true
+	}
+
+	// Check less than.
+	ltOp, ok := cmpOps[cmpArgs{LT, lType, rType}]
+	if !ok {
+		return 0, false
+	}
+	lt, err := ltOp.fn(EvalContext{}, l, r)
+	if err != nil {
+		panic(err)
+	}
+	if lt {
+		return -1, true
+	}
+	return 1, true
 }

--- a/sql/testdata/select_index
+++ b/sql/testdata/select_index
@@ -287,3 +287,8 @@ query ITT
 EXPLAIN SELECT a FROM t WHERE (b, c) = (5.0, 1)
 ----
 0 scan t@bc -
+
+query ITT
+EXPLAIN SELECT a FROM t WHERE b IN (5.0, 1)
+----
+0 scan t@b_desc -

--- a/sql/testdata/where
+++ b/sql/testdata/where
@@ -57,3 +57,11 @@ SELECT 'hello' SIMILAR TO v FROM kvString WHERE k SIMILAR TO 'like[1-2]'
 ----
 true
 false
+
+# Test mixed type tuple comparison.
+
+query II
+SELECT * FROM kv WHERE k IN (1, 5.0, 9)
+----
+1 2
+5 6

